cmd/dep/init.go

```diff
@@ -196,31 +196,31 @@ func (cmd *initCommand) Run(ctx *dep.Ctx, args []string) error {
 	// to generate the final lock memo.
 	s, err = gps.Prepare(params, sm)
 	if err != nil {
-		return errors.Wrap(err, "prepare solver")
+		return errors.Wrap(err, "init failed: unable to recalculate the lock digest")
 	}
 
 	p.Lock.SolveMeta.InputsDigest = s.HashInputs()
 
 	// Pass timestamp (yyyyMMddHHmmss format) as suffix to backup name.
 	vendorbak, err := dep.BackupVendor(vpath, time.Now().Format("20060102150405"))
 	if err != nil {
-		return err
+		return errors.Wrap(err, "init failed: first backup vendor/, delete it, and then retry the previous command: failed to backup existing vendor directory")
```

Comment:
I still feel this should be
```
<error message> : <recovery/help instructions>
```
Which would eventually look like:
```
<command> failed: <reason for failure>: <help instructions>
```

Also, instead of prefixing all the error messages with `<command> failed` inline, how about transferring this responsibility to [cmd/dep/main.go](https://github.com/golang/dep/blob/v0.3.1/cmd/dep/main.go?utf8=%E2%9C%93#L160).
```go
errLogger.Printf("%s failed: %v\n", cmd.Name(), err)
```
And this would take care of all other commands too :)

Reply:
Great idea, I love it!

Reply:
Do you have a suggestion for how to achieve the format you are suggesting? When we use `errors.Wrapf(err, "help instructions")` the result is: `help instructions: reason for failure`, so anything at the top of the stack is read first, and lower stack messages are last. So if we use it normally, it will look like `command failed: help instructions: reason for failure`.

Reply:
ah! right. I feel a need to change the return signature of `command.Run()`. Along with `error`, returning a help message would be nice.
```
Run(*dep.Ctx, []string) (string, error)
```
And could be used in main.go to print the help instruction after the error message. And in this way, errors would remain error and we don't have to mix errors with help messages.

What do you think?

Reply:
I see this improving how we print errors and help text at other places too. Like in `status`, right now we log all the helpful info to `stderr` and then return the error. As a result, the actual error is shown in the last line.
```
$ dep status
...
The status of 3 projects are unknown due to errors. Rerun with `-v` flag to see details.
failed to fetch updates
```
I feel it would be better to have the same presented as:
```
$ dep status
...
status failed: failed to fetch updates
The status of 3 projects are unknown due to errors. Rerun with `-v` flag to see details.
```

So that's
```
<command> failed: <reason>
<help message>
```
We can unify this format across all the commands if we change the return signature of `command.Run()`.

Reply:
I would prefer that we keep errors simple and handle them appropriately by each command.

Errors should just state that something wrong has happened. And it's up to the caller to handle it as it sees fit. This could mean either an action is taken or a helpful instruction is printed to the end user. This gives more control instead of jamming the instructions in the error message itself and forces more defined errors instead of relying on plain strings.

Reply:
How moving any more interesting refactoring of how commands handle errors, and print user instructions are in a separate issue. This was originally just for adding missing wrap's and ensuring that when the error is printed, it is with an appropriate amount of context.